### PR TITLE
Fix borrowck compiler errors for upvars contain "spurious" dereferences

### LINKED ISSUE #46003: borrowck compiler errors for upvars contain "spurious" dereferences
## STR

```Rust
fn main() {
    let mut y = 0;
    || {
        let b = &y;
        y += 1;
        *b
    };
}
```

## Result

```
error[E0506]: cannot assign to `**y` because it is borrowed
 --> src/main.rs:5:9
  |
4 |         let b = &y;
  |                  - borrow of `**y` occurs here
5 |         y += 1;
  |         ^^^^^^ assignment to borrowed `**y` occurs here

error: aborting due to previous error
```

`y` is an integer, so `**y` isn't actually a thing. We emit the dereferences because there are layers of indirection before we get to the upvar, but they aren't actually a part of the user-visible syntax.

### PR DESCRIPTION
**Pull request by vramana**
Fixes #46003

### CODE REVIEW

**Review by nikomatsakis — APPROVED** [review]

File: src/librustc_mir/borrow_check.rs [hunk 1]
```diff
@@ -1596,26 +1621,36 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+                                buf.push_str( &format!("*{}", &name));
+                            }
                         } else {
-                            buf.push_str(&"(*");
-                            self.append_lvalue_to_string(&proj.base, buf, Some(autoderef));
-                            buf.push_str(&")");
+                            if autoderef {
+                                self.append_lvalue_to_string(&proj.base, buf, Some(autoderef));
+                            } else {
+                                buf.push_str(&"(*");
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check.rs]: It's a separate bug (https://github.com/rust-lang/rust/issues/45185), but I'm pretty sure the `()` here are not necessary. That said, there is a fix for this in https://github.com/rust-lang/rust/pull/46040 (which will obviously conflict with this PR).

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@bors r+

We'll let the paren thing sort itself out in rebasing. =)

**Comment by nikomatsakis** [comment]
Needs rebase. =)

**Comment by vramana** [comment]
@nikomatsakis Done.

**Comment by vramana** [comment]
@nikomatsakis Seems like an unrelated build failure.

**Comment by arielb1** [comment]
@vramana 

You accidentally pushed a submodule update, please fix that

**Pull request merged**